File: frontend/src/types/ideaBuilder.ts
import type { ChatMessage } from './chat';
import type { Note } from './note';
import type { DraftTask } from './task';

export type IdeaBuilderPhase =
  | 'idle'
  | 'background'
  | 'resources'
  | 'familiarity'
  | 'existing_idea'
  | 'direction_choice'
  | 'refinement'
  | 'task_check'
  | 'completed';

export type IdeaDirectionChoice = 'migration' | 'improvement' | 'gap';
export type IdeaBuilderAction = 'generate_tasks' | 'keep_refining' | 'pause';

export type IdeaBuilderState = {
  workflow_id: number;
  phase: IdeaBuilderPhase;
  maturity_score: number;
  direction_choice: IdeaDirectionChoice | null;
  can_generate_tasks: boolean;
  memory_note_id: number | null;
  task_generation_status: 'not_requested' | 'drafted' | 'generated';
};

export type IdeaBuilderRunResponse = {
  workflow_id: number;
  phase: IdeaBuilderPhase;
  state: IdeaBuilderState;
  user_message: ChatMessage | null;
  assistant_messages: ChatMessage[];
  task_drafts: DraftTask[];
  memory_note: Note | null;
};

export type IdeaBuilderRespondPayload = {
  workflow_id: number;
  user_message?: string | null;
  direction_choice?: IdeaDirectionChoice | null;
  action?: IdeaBuilderAction | null;
};

File: backend/app/schemas/idea_builder.py
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.chat import ChatMessageRead
from app.schemas.note import NoteRead
from app.schemas.task import TaskPriority, TaskStatus


IdeaBuilderPhase = Literal[
    "idle",
    "background",
    "resources",
    "familiarity",
    "existing_idea",
    "direction_choice",
    "refinement",
    "task_check",
    "completed",
]
IdeaDirectionChoice = Literal["migration", "improvement", "gap"]
IdeaBuilderAction = Literal["generate_tasks", "keep_refining", "pause"]


class IdeaTaskDraftRead(BaseModel):
    title: str
    description: str = ""
    related_claim: str = ""
    owner: str = ""
    priority: TaskPriority = "medium"
    status: TaskStatus = "todo"


class IdeaBuilderStateRead(BaseModel):
    workflow_id: int
    phase: IdeaBuilderPhase = "idle"
    maturity_score: float = 0.0
    direction_choice: IdeaDirectionChoice | None = None
    can_generate_tasks: bool = False
    memory_note_id: int | None = None
    task_generation_status: Literal["not_requested", "drafted", "generated"] = "not_requested"


class IdeaBuilderStartRequest(BaseModel):
    workflow_id: int
    restart: bool = False


class IdeaBuilderRespondRequest(BaseModel):
    workflow_id: int
    user_message: str | None = Field(default=None, min_length=1)
    direction_choice: IdeaDirectionChoice | None = None
    action: IdeaBuilderAction | None = None


class IdeaBuilderRunResponse(BaseModel):
    workflow_id: int
    phase: IdeaBuilderPhase
    state: IdeaBuilderStateRead
    user_message: ChatMessageRead | None = None
    assistant_messages: list[ChatMessageRead] = []
    task_drafts: list[IdeaTaskDraftRead] = []
    memory_note: NoteRead | None = None
